samcli/lib/intrinsic_resolver/intrinsics_symbol_table.py

```diff
@@ -0,0 +1,419 @@
+        "eu-west-1": ["eu-west-1a", "eu-west-1b", "eu-west-1c"],
+        "eu-west-2": ["eu-west-2a", "eu-west-2b", "eu-west-2c"],
+        "eu-west-3": ["eu-west-3a", "eu-west-3b", "eu-west-3c"],
+        "cn-north-1": [],
+        "us-gov-west-1": [],
+    }
+
+    DEFAULT_PARTITION = "aws"
+    GOV_PARTITION = "aws-us-gov"
+    CHINA_PARTITION = "aws-cn"
+    CHINA_PREFIX = "cn"
+    GOV_PREFIX = "gov"
+
+    CHINA_URL_PREFIX = "amazonaws.com.cn"
+    DEFAULT_URL_PREFIX = "amazonaws.com"
+
+    AWS_NOTIFICATION_SERVICE_NAME = "sns"
+    ARN_SUFFIX = ".Arn"
+
+    CFN_RESOURCE_TYPE = "Type"
+
+    def __init__(
+        self,
+        template=None,
+        logical_id_translator=None,
+        default_type_resolver=None,
+        common_attribute_resolver=None,
+    ):
+        """
+        Initializes the Intrinsic Symbol Table so that runtime attributes can be resolved.
+
+        The code is defaulted in the following order logical_id_translator => parameters => default_type_resolver =>
+        common_attribute_resolver
+
+        If the item is a pseudo type, it will run through the logical_id_translator and if it doesn't exist there
+        it will generate a default one and save it in the logical_id_translator as a cache for future computation.
+        Parameters
+        ------------
+        logical_id_translator: dict
+            This will act as the default symbol table resolver. The resolver will first check if the attribute is
+            explicitly defined in this dictionary and do the relevant translation.
+
+            All Logical Ids and Pseudo types can be included here.
+            {
+                "RestApi.Test": {  # this could be used with RestApi.Deployment => NewRestApi
+                    "Ref": "NewRestApi"
+                },
+                "LambdaFunction": {
+                    "Ref": "LambdaFunction",
+                    "Arn": "MyArn"
+                }
+                "AWS::Region": "us-east-1"
+            }
+        default_type_resolver: dict
+            This can be used provide common attributes that are true across all objects of a certain type.
+            This can be in the format of
+            {
+                "AWS::ApiGateway::RestApi": {
+                    "RootResourceId": "/"
+                }
+            }
+            or can also be a function that takes in (logical_id, attribute_type) => string
+            {
+                "AWS::ApiGateway::RestApi": {
+                    "RootResourceId": (lambda l, a, p, r: p.get("ResourceId"))
+                }
+            }
+        common_attribute_resolver: dict
+            This is a clean way of specifying common attributes across all types.
+            The value can either be a function of the form string or (logical_id) => string
+            {
+                "Ref": lambda p,r: "",
+                "Arn:": arn_resolver
+            }
+        """
+        if template is None:
```

Comment:
You don't do anything be set template to `self._template` so this check can be wrapped into that assignment.